arangod/IResearch/IResearchLink.cpp

```diff
@@ -521,99 +536,78 @@ arangodb::Result IResearchLink::commitUnsafe() {
   if (!engine) {
     return {
       TRI_ERROR_INTERNAL,
-      "failure to get storage engine while committing arangosearch link '" + std::to_string(id()) + "'"};
+      "failure to get storage engine while committing arangosearch link '" + std::to_string(id()) + "'"
+    };
+  }
+
+  auto subscription = std::static_pointer_cast<IResearchFlushSubscription>(_flushSubscription);
+
+  if (!subscription) {
+    // already released
+    return {};
   }
 
   try {
-    // upcoming 'index_writer::commit()' will wait until all 'document_context's
-    // held by any transaction will be released, so 'tick' value can be less
-    // than actual engine tick when iresearch commit happens
-    auto const tick = engine->currentTick();
+    auto const lastTickBeforeCommit = engine->currentTick();
+
+    TRY_SCOPED_LOCK_NAMED(_commitMutex, commitLock);
+
+    if (!commitLock.owns_lock()) {
+      // commit is in progress
+      return {};
+    }
+
+    auto const lastCommittedTick = _lastCommittedTick;
 
-    _dataStore._writer->commit();
+    try {
+      // _lastCommittedTick is being updated in '_before_commit'
+      _dataStore._writer->commit(_before_commit);
+    } catch (...) {
+      // restore last committed tick in case of any error
+      _lastCommittedTick = lastCommittedTick;
+      throw;
+    }
 
-    SCOPED_LOCK(_readerMutex);
-    auto reader = _dataStore._reader.reopen(); // update reader
+    // get new reader
+    auto reader = _dataStore._reader.reopen();
 
     if (!reader) {
       // nothing more to do
-      LOG_TOPIC("37bcf", WARN, arangodb::iresearch::TOPIC)
+      LOG_TOPIC("37bcf", WARN, iresearch::TOPIC)
         << "failed to update snapshot after commit, run id '" << size_t(&runId)
         << "', reuse the existing snapshot for arangosearch link '" << id() << "'";
 
       return {};
```

Comment:
mmh, now it only logs a warning but returns an `ok()` Result from here. Shouldn't it return an error?